beep/features/featurizer_helpers.py

```diff
@@ -397,174 +397,124 @@ def get_hppc_ocv(processed_cycler_run, diag_pos):
-    cycles = hppc_cycle.cycle_index.unique()
-    diag_num = cycles[diag_pos]
-
-    selected_diag_df = hppc_cycle.loc[hppc_cycle.cycle_index == diag_num]
-    selected_diag_df = selected_diag_df.sort_values(by="test_time")
-    selected_diag_df["diagnostic_time"] = (selected_diag_df.test_time - selected_diag_df.test_time.min()) / 3600
-
-    return selected_diag_df
-
-
-def res_calc(selected_diag_df, steps, soc, step_ocv, step_cur, index):
-    """
-    This function calculates resistances at different socs and a specific pulse duration for a specified hppc cycle.
-
+    This function calculates resistance based on different socs and differnet time scales in hppc cycles.
+    
     Args:
-        selected_diag_df(pd.DataFrame): a dataframe for a specific diagnostic cycle you are interested in.
-        steps (list): list of step numbers for the specific occurrence of the diagnostic
-        if rpt_0.2C, occurs at cycle_index = [2, 37, 142, 244 ...], <diag_pos>=0 would correspond to cycle_index 2
-        soc (int): step index counter corresponding to the soc window of interest.
-        step_ocv (int): 0 corresponds to the 1h-rest, and 2 corresponds to the 40s-rest.
-        step_cur (int): 1 is for discharge, and 3 is for charge.
-        index (float or str): this will input a time scale for resistance (unit is second), e.g. 0.01, 5 or
-        'last' which is the entire pulse duration.
-
+        chosen(pd.DataFrame): a dataframe for a specific diagnostic cycle you are interested in.
+        soc (int): step index counter corresponding to the soc window of interest - 0, 1, 2, 3, 4 ... 
+        r_type (str): a string that indicates the time scale of the resistance you are calculating, e.g. 
+        'r_c_0s', 'r_c_3s', 'r_c_end', 'r_d_0s', 'r_d_3s', 'r_d_end'
+    
     Returns:
-        (a number) resistance at a specific soc in hppc cycles
+        charge/discharge resistance value (float) at this specific soc and time scale in hppc cycles
     """
+    
 
+    steps = chosen.step_index.unique()[1:6]
+    
     counters = []
-
+        
     for step in steps:
-        counters.append(
-            selected_diag_df[selected_diag_df.step_index == step].step_index_counter.unique().tolist()
-        )
-
-    if (len(counters[step_ocv]) - 1) < soc:
+        counters.append(chosen[chosen.step_index == step].step_index_counter.unique().tolist())
+        
+    # for charge 
+    if r_type[2] == 'c':
+        # 40 s short rest 
+        step_ocv = 2
+        step_cur = 3
+    
+    # for discharge 
+    if r_type[2] == 'd':
+        # one hour long rest 
+        step_ocv = 0
+        step_cur = 1
+    
+    # if there is no dataframe for ocv or step cur, it means this step is skipped, so return None directly 
+    try:
+        chosen_step_ocv = chosen[(chosen.step_index_counter == counters[step_ocv][soc])]
+        chosen_step_cur = chosen[chosen.step_index_counter == counters[step_cur][soc]]
+    except:
```

Comment:
Its usually best practice to catch a specific type of exception that you expect. For instance, here you might be expecting a `KeyError` or something similar, based on the comment above. btw, thank you for adding the comment above since it makes it much clearer what is going on.

So it would look like
except KeyError:

Reply:
Thank you Patrick! I added in except "IndexError"